bpfd-operator/controllers/bpfd-agent.go

```diff
@@ -0,0 +1,504 @@
+			Reason:  "bpfdLoaded",
+			Message: "Successfully loaded ebpfProgram",
+		}
+
+		updateStatusFunc(loadedCondition)
+
+		l.Info("Program loaded via bpfd", "bpfd-program-uuid", uuid)
+		return false, nil
+	}
+
+	// This function unloads the bpf program via bpfd and removes the bpfd-agent
+	// finalizer from the ebpfProgram Object
+	unloadFunc := func() (bool, error) {
+		for uuid := range ebpfProgram.Spec.ProgramMap {
+			unloadRequest, err := internal.BuildBpfdUnloadRequest(uuid)
+			if err != nil {
+				// Add a condition and exit do requeue, bpfd might become ready
+				return true, fmt.Errorf("failed to generate bpfd unload request: %v",
+					err)
+			}
+
+			err = r.unloadBpfdProgram(ctx, unloadRequest)
+			if err != nil {
+				failUnloadCondition := metav1.Condition{
+					Type:    "NotUnloaded",
+					Status:  metav1.ConditionTrue,
+					Reason:  "bpfdNotUnloaded",
+					Message: "Failed to unload ebpfProgram",
+				}
+
+				updateStatusFunc(failUnloadCondition)
+
+				return true, fmt.Errorf("failed to unload ebpfProgram via bpfd: %v",
+					err)
+			}
+
+			l.Info("Program loaded via bpfd", "bpfd-program-uuid", uuid)
+		}
+
+		return false, nil
+	}
+
+	// TODO(astoycos) This will need to end up being a list of loadRequests
+	// if a given ebpfProgramConfig selects more than one attach point
+	// (i.e if we support a pod LabelSelector for interfaces) For now
+	// we only support specifying a single node interface in the API so
+	// there will only be a single loadRequest per ebpfProgramConfig Object.
+	loadRequest, err := internal.BuildBpfdLoadRequest(ebpfProgramConfig)
+	if err != nil {
+		return true, fmt.Errorf("failed to generate bpfd load request: %v",
+			err)
+	}
+
+	// Compare the desired state to existing node state
+	v, ok := nodeState[ebpfProgramConfig.Spec.Name]
+	// ebpfProgram doesn't exist on node
+	if !ok {
+		// Make sure if we're not selected just exit
+		if !isNodeSelected {
+			// Write NodeNodeSelected status
+			nodeNotSelectedCondition := metav1.Condition{
+				Type:    "NotSelected",
+				Status:  metav1.ConditionTrue,
+				Reason:  "nodeNotSelected",
+				Message: "This node is not selected to run the ebpfProgram",
+			}
+			updateStatusFunc(nodeNotSelectedCondition)
+
+			return false, nil
+		}
+
+		// If EbpfProgramConfig is being deleted just remove agent finalizer so the
+		// owner relationship can take care of cleanup
+		if !ebpfProgramConfig.DeletionTimestamp.IsZero() {
+			if controllerutil.ContainsFinalizer(ebpfProgram, bpfdAgentFinalizer) {
+				controllerutil.RemoveFinalizer(ebpfProgram, bpfdAgentFinalizer)
+				err := r.Update(ctx, ebpfProgram)
+				if err != nil {
+					return false, err
+				}
+			}
+
+			return false, nil
+		}
+
+		// otherwise load it
+		return loadFunc(loadRequest)
+	}
+
+	// EbpfProgram exists but either EbpfProgramConfig is being deleted or node is no
+	// longer selected....unload program
+	if !ebpfProgramConfig.DeletionTimestamp.IsZero() || !isNodeSelected {
```

Comment:
Seems you are doing a bit of work if the ebpfProgram does not exist (i.e. "Always create the ebpfProgram Object if it doesn't exist" and building a loadRequest object). is it worth adding a check earlier to just bail if the ebpfProgram does not exist and node is not Selected or ebpfProgramConfig is being deleted?

Reply:
So we actually always do want to create the ebpfProgram Object even if our node isn't selected, since then in the status we can report something like 

```
	// Write NodeNodeSelected status
	nodeNotSelectedCondition := metav1.Condition{
		Type:    string(EbpfProgCondNotSelected),
		Status:  metav1.ConditionTrue,
		Reason:  "nodeNotSelected",
		Message: "This node is not selected to run the ebpfProgram",
	}
```

> is it worth adding a check earlier to just bail if the ebpfProgram does not exist and node is not Selected or ebpfProgramConfig is being deleted?

Ehh I'm worried about races here.... If we rely on "the ebpfProgram Object existing" to ALWAYS mean that program is actually loaded on the node things get a little messy... The way I have things now we explicitly check if that program exists via BPFD (`nodeState` map) before doing anything which seems to cut down the number of corner cases